File: content/02-two-pointers/trapping_rain_water.js
/**
 * Problem: Trapping Rain Water
 * GeeksforGeeks: https://practice.geeksforgeeks.org/problems/trapping-rain-water/0
 */

function trap(height) {
    /**
     * Time Complexity: O(N)
     * Space Complexity: O(1)
     */
    if (!height || height.length === 0) return 0;
    
    let left = 0, right = height.length - 1;
    let leftMax = height[left], rightMax = height[right];
    let water = 0;
    
    while (left < right) {
        if (leftMax < rightMax) {
            left++;
            leftMax = Math.max(leftMax, height[left]);
            water += leftMax - height[left];
        } else {
            right--;
            rightMax = Math.max(rightMax, height[right]);
            water += rightMax - height[right];
        }
    }
    
    return water;
}

if (require.main === module) {
    const assert = require('assert');
    assert.strictEqual(trap([0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]), 6);
    assert.strictEqual(trap([4, 2, 0, 3, 2, 5]), 9);
    console.log("JS: trap passed tests!");
}

File: content/02-two-pointers/trapping_rain_water.py
"""
Problem: Trapping Rain Water
GeeksforGeeks: https://practice.geeksforgeeks.org/problems/trapping-rain-water/0
"""

def trap(height: list[int]) -> int:
    """
    Time Complexity: O(N)
    Space Complexity: O(1)
    """
    if not height:
        return 0
        
    left, right = 0, len(height) - 1
    left_max, right_max = height[left], height[right]
    water = 0
    
    while left < right:
        if left_max < right_max:
            left += 1
            left_max = max(left_max, height[left])
            water += left_max - height[left]
        else:
            right -= 1
            right_max = max(right_max, height[right])
            water += right_max - height[right]
            
    return water

if __name__ == "__main__":
    assert trap([0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]) == 6
    assert trap([4, 2, 0, 3, 2, 5]) == 9
    print("Python: trap passed tests!")
